backend/src/akvo/flow_services/uploader.clj

```diff
@@ -287,44 +293,57 @@
+           (add-image-upload-messages bucket file-name form-instance-id)))))
 
 (defn- get-image-questions
   [xml-form]
   (filter #(= (.toLowerCase (.getType %)) "photo") (mapcat #(.getQuestion %) (.getQuestionGroup xml-form))))
 
 (defn- get-form-instance
   "Retrieve form to which all images will be bulk uploaded"
-  [app-id instance-id-str]
+  [app-id form-instance-id]
   (gae/with-datastore [ds (util/datastore-spec app-id)]
-                      (try
-                        (let [instance-id (Long/parseLong instance-id-str)]
-                          (query/entity ds "SurveyInstance" instance-id))
-                        (catch NumberFormatException e
-                          (errorf "\"%s\" is not a valid form instance id" instance-id-str)))))
+                      (query/entity ds "SurveyInstance" form-instance-id)))
+
+(defn- map-folder-data
+  [folder]
+  (let [grouped-files (group-by #(fs/base-name (fs/parent %)) (get-images folder))]
+    (mapv (fn [[folder-name images]]
+            (try
+              (let [form-instance-id (Long/parseLong folder-name)]
+                {:form-instance-id form-instance-id
+                 :images images})
+              (catch NumberFormatException e
+                {:error true
+                 :folder-name folder-name})))
+          grouped-files)))
 
 (defn bulk-image-upload
   "Prepare uploaded images to be pushed to the flow backend"
   [base-url unique-identifier file-name upload-domain]
   (let [app-id (get @config/s3bucket->app-id (config/get-bucket-name upload-domain))
+        bucket (:s3bucket (config/find-config app-id))
         path (format "%s/%s" (get-path) unique-identifier)
         _ (combine path file-name)
         _ (cleanup path)
         unzipped-dir (unzip-file path file-name)
-        files-by-folder (group-by #(fs/base-name (fs/parent %)) (get-images unzipped-dir))
-        form-instance (get-form-instance app-id (key (first files-by-folder)))
-        form-id (.getProperty form-instance "surveyId")
-        xml-form (util/get-published-form app-id form-id)
-        image-questions (get-image-questions xml-form)]
-    (->> files-by-folder
-         (mapv (fn [[form-instance-id images]]
-                 (let [folder-data {:form-instance-id (Long/parseLong form-instance-id)
-                                    :images images}]
-                   (process-image-upload-folder app-id base-url file-name folder-data image-questions)))))))
+        folder-data (map-folder-data unzipped-dir)
+        form-instance-id (some :form-instance-id folder-data)
+        form-instance (when form-instance-id (get-form-instance app-id form-instance-id))
+        form-id (when form-instance (.getProperty form-instance "surveyId"))
+        xml-form (when form-id (util/get-published-form app-id form-id))
+        image-questions (when xml-form (get-image-questions xml-form))]
+    (if xml-form
+      (->> folder-data
+           (mapv (fn [folder]
```

Comment:
Not sure if I understand this code, but `map` and `mapv` is when you want to return *values* to the caller. It seems to me that we want to *process* a collection of things? 🤔  If that is the case `doseq` is more appropiate

Reply:
I had `doseq` originaly but Juan seemed to think mapv was more appropriate. 😄 

https://github.com/akvo/akvo-flow-services/blob/02b5874c86036bd15185add7ff443319cf140a62/backend/src/akvo/flow_services/uploader.clj#L327